src/builder/create_attachment.rs

```diff
@@ -117,4 +106,174 @@ impl CreateAttachment {
+///     EditAttachments::new().keep(msg.attachments[0].id)
+/// )).await?;
+/// # Ok(()) }
+/// ```
+///
+/// ## Delete only the first attachment
+///
+/// ```rust,no_run
+/// # use serenity::all::*;
+/// # async fn _foo(ctx: Context, mut msg: Message, my_attachment: CreateAttachment) -> Result<(), Error> {
+/// msg.edit(ctx, EditMessage::new().attachments(
+///     EditAttachments::new_keep_all(&msg).dont_keep(msg.attachments[0].id)
+/// )).await?;
+/// # Ok(()) }
+/// ```
+///
+/// # Notes
+///
+/// Internally, this type is used not just for message editing endpoints, but also for message
+/// creation endpoints.
+#[derive(Default, Debug, Clone, serde::Serialize)]
+#[serde(transparent)]
+#[must_use]
+pub struct EditAttachments {
+    new_and_existing_attachments: Vec<NewOrExisting>,
+}
+
+impl EditAttachments {
+    /// An empty attachments builder.
+    ///
+    /// Existing attachments are not kept by default, either. See [`Self::new_keep_all()`] or
+    /// [`Self::keep()`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Creates a new attachments builder that keeps all existing attachments.
+    ///
+    /// Shorthand for [`Self::new()`] and calling [`Self::keep()`] for every
+    /// [`AttachmentId`] in [`Message::attachments`].
+    ///
+    /// If you only want to keep a subset of attachments from the message, either implement this
+    /// method manually, or use [`Self::dont_keep()`].
+    ///
+    /// **Note: this EditAttachments must be run on the same message as is supplied here, or else
+    /// Discord will throw an error!**
+    pub fn new_keep_all(msg: &Message) -> Self {
+        Self {
+            new_and_existing_attachments: msg
+                .attachments
+                .iter()
+                .map(|a| {
+                    NewOrExisting::Existing(ExistingAttachment {
+                        id: a.id,
+                    })
+                })
+                .collect(),
+        }
+    }
+
+    /// This method adds an existing attachment to the list of attachments that are kept after
+    /// editing.
+    pub fn keep(mut self, id: AttachmentId) -> Self {
+        self.new_and_existing_attachments.push(NewOrExisting::Existing(ExistingAttachment {
+            id,
+        }));
+        self
+    }
+
+    /// This method removes an existing attachment from the list of attachments that are kept after
+    /// editing.
+    pub fn dont_keep(mut self, id: AttachmentId) -> Self {
+        #[allow(clippy::match_like_matches_macro)] // `matches!` is less clear here
+        self.new_and_existing_attachments.retain(|a| match a {
+            NewOrExisting::Existing(a) if a.id == id => false,
+            _ => true,
+        });
+        self
+    }
+
+    /// Adds a new attachment to the attachment list.
+    #[allow(clippy::should_implement_trait)] // Clippy thinks add == std::ops::Add::add
+    pub fn add(mut self, attachment: CreateAttachment) -> Self {
```

Comment:
Can't think of an alternative name, since `push` doesn't really make sense. Willing to discuss further though.

Reply:
Why an alternative name? I like `add` very much - short, simple, self-explanatory. The fact that clippy throws a fit is not a reason to discard that name in my opinion

Reply:
Mainly because it's slightly ambiguous what you're adding if you only look at the method name, if it's existing or new attachments.

Reply:
In another comment I explained my envisioned mental model here. The idea is that existing attachments shouldn't be thought of as a pushable list, but a set with togglable items. Hence, semantically, the only pushable list is the list of new attachments and `add` is unambiguous